### DISCUSSION TITLE
Establish a first-time contributor props culture in Gutenberg

### DISCUSSION THREAD

**Discussion by priethor** [WordPress/gutenberg#38244] [General]
## What problem does this address?

WordPress has a long history of giving props to first-time contributors in each release and keeping track of new folks joining the project. However, Gutenberg has only recently started acknowledging first-time contributors in [Slack meetings](https://wordpress.slack.com/archives/C02QB2JS7/p1642604332351300) and [Gutenberg release posts](https://make.wordpress.org/core/2022/01/19/whats-new-in-gutenberg-12-4-19-january/#props-to-new-contributors). Currently, it is a semi-manual process done [by checking the PRs merged](https://github.com/WordPress/gutenberg/pulls?q=is%3Amerged+is%3Apr+label%3A%22First-time+Contributor%22+milestone%3A%22Gutenberg+12.5%22) in the last release cycle labeled as `First-time contributor`, but needs to be double-checked and doesn't include contributions other than PRs.

## What is your proposed solution?

The props process in Gutenberg could use some refining and automation, both to ease the process and to give credit to non-PR contributions as well. Some improvements include:
- [ ] Automatically generate the list of first-time contributors when generating the changelog
- [ ] Create a label `First-time reporter` and `First-time commenter` to track non-PR contributions as well.
- [ ] Update the release process documentation to include the props section in the release post.

What other relevant contributions should be highlighted and celebrated? What mechanisms can we set to ensure we catch them all?

**Comment by getdave** [comment]
Thank you for raising this @priethor. I agree it's important. I appreciate that automating this process will be tricky but I believe it will be a good first step towards improving on what we have now. We can then iterate towards better tooling, systems and cultural changes which will improve the situation still further.

To this end I've raised [this PR which tweaks the changelog script to generate a list of FTC](https://github.com/WordPress/gutenberg/pull/38372)s. Let me know what you think.

**Comment by desrosj** [comment]
This is great!

In the broader context of giving better and more accurate attribution through the "Props" used to compile the credits API for each release, I've published the following proposal for adjustments in order to better track contribution efforts: https://make.wordpress.org/core/2022/01/31/ensuring-proper-attribution-for-contributions-to-wordpress-on-github/.

**Comment by getdave** [comment]
> Create a label First-time reporter and First-time commenter to track non-PR contributions as well.

@priethor If you want to go ahead and do this we could extend https://github.com/WordPress/gutenberg/pull/38372 to include these props as well. By default nothing would be added _if_ there are no PRs/Issues found containing those labels.

**Reply by priethor** [reply]
I think it could be a good next step. Automating the `First-time reporter` label should be fairly easy, but I am not sure about the comment one.

**Comment by bph** [comment]
Another aspect is to add Gutenberg GitHub stats to @audrasjb This week in Core posts. Right now they only include tract numbers. 
My thinking on the whole topic: 

If we can iterate on the contributors list on a weekly basis, we can get to a point when a big WordPress release comes around, there might be a streamlined process available. At least that's a hope I have. 

This week's numbers from this repository would be using the "Insight" tab: 
![Screen Shot 2022-02-07 at 10 22 53](https://user-images.githubusercontent.com/39980/152817214-1ef43d73-9e6a-4e96-a175-9786f7e91846.png)

100 commits
67 contributors (code only) 
92 new issues
69 close issues

- Do we only want to match Trac's numbers with GitHub equivalents? 

JB posts a list of all contributors and the number of commits 
![Screen Shot 2022-02-02 at 16 50 00](https://user-images.githubusercontent.com/39980/152818318-fd215aed-66c4-4b1b-85a6-2e4deef8934b.png)

Is it possible to create a similar list from GitHub contributors? 

![Screen Shot 2022-02-02 at 16 53 06](https://user-images.githubusercontent.com/39980/152818511-70eb0e9c-8ac5-4675-a66c-9ffa99b727e4.png)

Are we able to identify First contributors on GitHub repo as @getdave did for the last plugin release on a weekly basis?

The bigger question, that was also raised in @desrosj proposal is what does the team count as contribution. It would be great if it can match the overall core approach. 

Quoteing from [Core Handbook > Best Practices > Commit Message > Props](https://make.wordpress.org/core/handbook/best-practices/commit-messages/#props)

> "Props should be given to all those who contributed to the final commit, whether through _patches, refreshed patches, code suggested otherwise, design, writing, user testing, or other significant investments of time and effort_. Usernames are parsed for the credits list and WordPress.org profiles.
> 
> In the case of bug reports, props should also be given to the _reporter of the bug._
> 
> Check any tickets which were closed as a duplicate in case they contain contributions that warrant props too."

Following these guildelines, The PR should list the usernames for props. 

How could this be implemented?

**Reply by getdave** [reply]
> Are we able to identify First contributors on GitHub repo as @getdave did for the last plugin release on a weekly basis?

Yes. This is [now part of the changelog creation script](https://github.com/WordPress/gutenberg/pull/38372). It is based on folks adding the "First time contributor" label to any PRs.

As [I suggested to @priethor](https://github.com/WordPress/gutenberg/discussions/38244#discussioncomment-2096579) we could extend this to also have labels for other types of contributions. 

That said there will still be other types of contribution that will be harder to automated - PR reviews are one that comes to mind. To serve more complex use cases we'd probably need to use some kind of freeform attribution within Github comments. Let's see if there's some kind of Open standard for "props" or "kudos" which we could use for this.

**Comment by bph** [comment]
In trac, the committer adds props to the issues/ticket
Example: ([trac 52665](https://core.trac.wordpress.org/changeset/52665))
![Screen Shot 2022-02-09 at 19 05 44](https://user-images.githubusercontent.com/39980/153311515-31ebf4d3-fe6a-48d9-a9bc-d763f71fa8e2.png)
`Props blogaid, linux4me2, mgol, Clorith.` - wporg usernames comma delimited, after the word `Props` in a single line

**Reply by getdave** [reply]
These props seem fairly blanket then? Not differentiation between why the props are provided. If we're ok with that then it's probably doable to parse these out.

When would these props be left? Just in any of the comments? Or the merge commit message description? Open to suggestions.

**Reply by bph** [reply]
Exactly. Just a line with the usernames. That's all information needed to give props. :-)

**Reply by bph** [reply]
>When would these props be left? Just in any of the comments? Or the merge commit message description? Open to suggestions.

Sorry didn't answer that yesterday. We have a few options, it depends who should be responsible to collect the props and where ti add them, so a script could collect them 

- Merge Commit Message > responsible?
- Request for review > responsible?

If this follows the practice on core trac, it could be  added to the original description of the PR by the person requesting final review before the merge. 

What do you think?

**Reply by getdave** [reply]
> Merge Commit Message

Seems good except that if it's forgotten it might be difficult to amend that retrospectively.

> Request for review

I wonder if it's possible to grab the usernames of users who leave reviews on a PR that has been merged 🤔

**Reply by bph** [reply]
I wonder that too. I have not seen much in the API. I also don't have any privileges on the GitHub repository.

If the PR closes/fixes an issue, the issue reporter and the people, who helped narrow things down to the cause should also get props.

**Comment by bph** [comment]
An example for @getdave list of contributors on [changelog on 12.6.RC1](https://github.com/WordPress/gutenberg/releases/tag/v12.6.0-rc.1) - This is fabulous.  https://github.com/WordPress/gutenberg/releases/tag/v12.6.0-rc.1
![Screen Shot 2022-02-10 at 08 18 31](https://user-images.githubusercontent.com/39980/153416924-599aa2a4-2bc3-42a7-8b3c-8c3d9e7d95c8.png)

Great list of first time contributors. The all contributors list usernames is truncated, though.
There have to be more contributors, than four for the whole release., though...

**Reply by getdave** [reply]
@bph I think we need to differentiate the current mechanic from what we want to achieve.

Currently the `First time contributors` list is generated based on any PRs with the label `First time contributor`. This seems to cause Github to _auto_-generate the `Contributors` list at the bottom of the changelog which is..._unintended_.

In the future what I think we are aiming for is something like:

* List of first time 
    - contributors.
    - reviewers.
    - Issue reporters.
* A general "contributors" list including everyone who's helped contribute to this release.

I understood the project wanted to congratulate first time contributors particularly. 

It would also be nice to list _all_ contributors separately.

That's just a suggestion. Open to more discussion. Thanks 🙇

**Reply by getdave** [reply]
@bph Here is [a new PR I just spun up which should generate a full "Contributors" section based on the usernames of those who merged PRs](https://github.com/WordPress/gutenberg/pull/38704).

**Reply by bph** [reply]
Yeah, you are right, @getdave . Seems in a very convoluted way,  I was dreaming a bit bigger. All three items entered my  attention in the same week. 

I see three goals, in the sense of while we are working on this, maybe we can also think about two additional very related things: 

1. The first contributors on a release. 
2. This Week in Core post and come up with a method/process to provide Gutenberg numbers and list of contributors to JB for his post. All contriubtors for the week, alternatively for the 14-day releases, separate out the first time contributors. 
3. Giving **all** contributors to Gutenberg props for the WordPress release, no only code committers. 

You already solved one - via the tag. So that additional information we can give every two weeks to JB with the info from the Insights. 

To get to Number 3 all contributors, the team needs to change slightly the PR process to add props.  I don't know if it needs to be handled on the merge process or before the review request. 

It's like a three tiered process :-)

**Reply by bph** [reply]
> @bph Here is [a new PR I just spun up which should generate a full "Contributors" section based on the usernames of those who merged PRs](https://github.com/WordPress/gutenberg/pull/38704).

Thank you for putting this together. Glad, you got it merged earlier today. Europe FTW!  How can I test it after the fact?

**Comment by bph** [comment]
Just saw that @desrosj  is working on a [Proof of Concept: Props Bot ](https://github.com/WordPress/gutenberg/pull/38538)

**Comment by bph** [comment]
Just a brief update: Gutenberg 12.7 Changelog now has the list of code contributors for this version.
<img width="904" alt="screenshot_2022-02-23_at_14 02 55" src="https://user-images.githubusercontent.com/39980/155344387-659c2749-eea2-4f8a-8e49-a6313d4273e9.png">

52 people merged PRs to the version. It's great to have this number and the list.

**Comment by getdave** [comment]
We should consider making use of https://github.com/10up/credits-generator. Kudos to @bph for the recommendation.

> This tiny package scans and generates a markdown command-separated list of GitHub users contributed to a repo including issue reporters, PR creators, and all commenters.

**Reply by priethor** [reply]
On the subject of a props generator, what do you think about using the [All Contributors Bot & Specification](https://allcontributors.org/)? Apart from being a pretty popular specification, it was [first suggested in Gutenberg](https://github.com/WordPress/gutenberg/issues/1223#issuecomment-309201610) back in 2017.

**Reply by getdave** [reply]
That looks interesting. I'm wondering if there's a way to config it to run against a given release only rather than "all time"? I seems like it's updating a constant file which references contributions from "all time".

I guess we could have a per release file which is updated and then reset at the time of each release.

**Reply by priethor** [reply]
> I seems like it's updating a constant file which references contributions from "all time".

Indeed, in this case, I was bringing up `All Contributors` as a way of kudo-ing all contributors somewhere else than during releases; for that purpose, the `credits-generator` tool proposed above seems a better fit.

**Reply by getdave** [reply]
Ah I see. So we're proposing a complimentary contributors log to sit in a `CONTRIBUTORS` file at the root of the README. Interesting.

**Reply by jeffpaul** [reply]
> On the subject of a props generator, what do you think about using the [All Contributors Bot & Specification](https://allcontributors.org/)? Apart from being a pretty popular specification, it was [first suggested in Gutenberg](https://github.com/WordPress/gutenberg/issues/1223#issuecomment-309201610) back in 2017.

Last I checked that out it was a good bit time-intensive to coordinate and the visuals were a bit too much for me but regardless I'm open to any process wherein we better credit ALL contributions.
